onnxruntime/contrib_ops/cpu/bert/qembed_layer_norm.cc

```diff
@@ -0,0 +1,243 @@
+
+  // Request outputs:
+  TensorShape output_shape({batch_size, sequence_length, hidden_size});
+  Tensor* output = context->Output(0, output_shape);
+
+  TensorShape mask_index_shape({batch_size});
+  Tensor* mask_index = context->Output(1, mask_index_shape);
+
+  // Grab pointers to buffers each Tensor represents:
+  const uint8_t* word_embedding_data = word_embedding->template Data<uint8_t>();
+  const uint8_t* position_embedding_data = position_embedding->template Data<uint8_t>();
+  const uint8_t* segment_embedding_data =
+      has_segment_embedding ? segment_embedding->template Data<uint8_t>() : nullptr;
+  const uint8_t* gamma_data = gamma->template Data<uint8_t>();
+  const uint8_t* beta_data = beta->template Data<uint8_t>();
+
+  T* output_data = output->template MutableData<T>();
+
+  // Perform the Op:
+  {
+    std::atomic_bool failed{false};
+
+    int n = batch_size * sequence_length;
+    concurrency::ThreadPool::TryBatchParallelFor(
+        context->GetOperatorThreadPool(), n, [=, &failed](ptrdiff_t index) {
+          int word_col_index = input_ids_data[index];
+          if (word_col_index < 0 || word_col_index >= word_embedding_length) {
+            failed.store(true, std::memory_order_release);
+            return;
+          }
+          int position_col_index = index % sequence_length;
+          if (position_col_index >= position_embedding_length) {
+            failed.store(true, std::memory_order_release);
+            return;
+          }
+          int segment_col_index = 0;
+          if (nullptr != segment_ids_data) {
+            segment_col_index = segment_ids_data[index];
+            if (segment_col_index < 0 || segment_col_index >= segment_embedding_length) {
+              failed.store(true, std::memory_order_release);
+              return;
+            }
+          }
+
+          // Grab inputs for the embeddings for the current batch index:
+          const uint8_t* input_word_embedding = word_embedding_data + word_col_index * hidden_size;
+          const uint8_t* input_position_embedding =
+              position_embedding_data + position_col_index * hidden_size;
+          const uint8_t* input_segment_embedding = nullptr;
+          if (segment_embedding_data != nullptr) {
+            input_segment_embedding = segment_embedding_data + segment_col_index * hidden_size;
+          }
+
+          T* output = output_data + (index * hidden_size);
+
+          T sum = static_cast<T>(0);
+          for (int i = 0; i < hidden_size; ++i) {
+            T subtotal = Dequantize<uint8_t>(input_word_embedding[i],
+                                             word_embedding_scale,
+                                             word_embedding_zero_point) +
+                         Dequantize<uint8_t>(input_position_embedding[i],
+                                             position_embedding_scale,
+                                             position_embedding_zero_point);
```

Comment:
This is slow. To improve the performance, you can use table query, an approach we use for other operators. 

As word_embedding_scale and word_embedding_zero_point are constant, you can create a hash table with key is input_word_embedding[i] and value is  (input_word_embedding[i] - word_embedding_zero_point ) * word_embedding_scale

Reply:
Even for the case word_embedding_scale and word_embedding_zero_point are non-constant, you can calculate the hash table ahead dynamically

Reply:
Further more, you can make the value as (input_word_embedding[i] - word_embedding_zero_point ) * word_embedding_scale + (input_segment_embedding[i] -  segment_embedding_zero_point) * segment_embedding_scale

Reply:
[https://github.com/microsoft/onnxruntime/blob/53d1d55ea88f9c8a7da74dd67608a75236db9e7b/onnxruntime/contrib_ops/cpu/qlinear_lookup_table.h](https://github.com/microsoft/onnxruntime/blob/53d1d55ea88f9c8a7da74dd67608a75236db9e7b/onnxruntime/contrib_ops/cpu/qlinear_lookup_table.h)

Reply:
This is great! I didn't know these things existed. Mind if I do this as a fast-follow? This PR is getting rather large. 

I was thinking that a follow up PR that used these techniques could use as a canonical PR with documentation to help future contributors write these optimizations for quantized kernels.

Reply:
Sounds good.